nixos/modules/security/tpm2.nix

```diff
@@ -0,0 +1,120 @@
+
+  # This snippet is taken from tpm2-tss/dist/tpm-udev.rules, but modified to allow custom user/groups
+  # The idea is that the tssUser is allowed to acess the TPM and kernel TPM resource manager, while
+  # the tssGroup is only allowed to access the kernel resource manager
+  # Therefore, if either of the two are null, the respective part isn't generated
+  udevRules = tssUser: tssGroup: ''
+    ${lib.optionalString (tssUser != null) ''KERNEL=="tpm[0-9]*", MODE="0660", OWNER="${tssUser}"''}
+    ${lib.optionalString (tssUser != null || tssGroup != null)
+      ''KERNEL=="tpmrm[0-9]*", MODE="0660"''
+      + lib.optionalString (tssUser != null) '', OWNER="${tssUser}"''
+      + lib.optionalString (tssGroup != null) '', GROUP="${tssGroup}"''
+     }
+  '';
+
+in {
+  options.security.tpm2 = {
+    enable = lib.mkEnableOption "Trusted Platform Module 2 support";
+
+    tssUser = lib.mkOption {
+      description = "Name of the tpm device-owner and service user, set if applyUdevRules is set";
+      type = lib.types.nullOr lib.types.str;
+      default = "tss";
+    };
+
+    tssGroup = lib.mkOption {
+      description = "Group of the tpm kernel resource manager (tpmrm) device-group, set if applyUdevRules is set";
+      type = lib.types.nullOr lib.types.str;
+      default = "tss";
+    };
+
+    applyUdevRules = lib.mkOption {
+      description = ''
+        Whether to make the /dev/tpm[0-9] devices accessible by the tssUser, or
+        the /dev/tpmrm[0-9] by tssGroup respectively
+      '';
+      type = lib.types.bool;
+      default = true;
+    };
+
+    abrmd = lib.mkOption {
+      description = "TPM2 userspace resource manager service";
+      type = lib.types.submodule ({
+        options = {
+          enable = lib.mkEnableOption "Trusted Platform 2 userspace resource manager daemon";
+
+          user = lib.mkOption {
+            description = ''
+              User tpm2-abrmd should run under. This user should have access to
+              the /dev/tpm[0-9] device.
+            '';
+            type = lib.types.str;
+            default = config.security.tpm2.tssUser;
+          };
+
+          package = lib.mkOption {
+            description = "tpm2-abrmd package to use";
+            type = lib.types.package;
+            default = pkgs.tpm2-abrmd;
+          };
+
+          setEnvironmentVariables = lib.mkOption {
+            description = ''
+              Set common environment variables to use the tpm2-abrmd resource manager
+              The variables are
+                - TPM2TOOLS_TCTI
+                - TPM2_PKCS11_TCTI
+            '';
+            type = lib.types.bool;
+            default = false;
+          };
+        };
+      });
+    };
+  };
+
+  config = lib.mkIf cfg.enable (lib.mkMerge [
+    {
+      services.udev.extraRules = lib.optionalString cfg.applyUdevRules (udevRules cfg.tssUser cfg.tssGroup);
+
+      users.users = lib.mkIf (cfg.tssUser != null) {
+        "${cfg.tssUser}" = {
```

Comment:
You could do this as well, same for `users.groups`
```nix
users.users.${cfg.tssUser} = lib.mkIf (cfg.tssUser != null) {
  isSystemUser = true;
}
```

However usually such `user`/`group` options are of type `types.str` without allowing `null`, and the check is done as `mkIf (cfg.tssUser == "tss")` (meaning, only create this user in the module when the default value is used). I'd recommend doing that here